File: ocaml/lib/kan/coyoneda.ml
module type Contravariant = sig
  type 'a t
  val contramap : ('b -> 'a) -> 'a t -> 'b t
end

module Coyoneda (F : Contravariant) = struct
  type 'a t = {
    f : 'a -> 'b;
    m : 'b F.t;
  }

  let coyoneda f m = { f; m }

  let contramap f { f = g; m } = { f = f @@ g; m }
end

File: rust/src/kan/coyoneda.rs
struct Coyoneda<F, A>
where
    F: Contravariant,
{
    f: Box<dyn Fn(A) -> F::Res>,
    m: F::Res,
}

impl<F, A> Coyoneda<F, A>
where
    F: Contravariant,
{
    fn new(f: Box<dyn Fn(A) -> F::Res>, m: F::Res) -> Self {
        Coyoneda { f, m }
    }

    fn contramap<B>(self, g: Box<dyn Fn(B) -> A>) -> Coyoneda<F, B> {
        Coyoneda {
            f: Box::new(move |x| self.f(g(x))),
            m: self.m,
        }
    }
}

trait Contravariant {
    type Res;
    fn contramap<A, B>(self, f: Box<dyn Fn(B) -> A>) -> Self::Res;
}